replica-common/src/main/java/com/graphhopper/OsmHelper.java

```diff
@@ -38,27 +61,101 @@ public int getSegmentIndexForGhEdge(int ghEdgeId) {
         return ghEdgeIdToSegmentIndexMapping.getInt(pointer);
     }
 
-    public static Map<String, String> getLanesTag(long osmId, Map<Long, Map<String, String>> osmIdToLaneTags) {
-        return osmIdToLaneTags.getOrDefault(osmId, null);
+    public static Map<String, String> getWayTag(long osmId, Map<Long, Map<String, String>> osmIdToWayTags) {
+        return osmIdToWayTags.getOrDefault(osmId, null);
     }
 
-    public static String getHighwayFromOsmWay(ReaderWay way) {
-        if (way.hasTag("highway")) {
-            return way.getTag("highway");
+    public static String getTagValueFromOsmElement(ReaderElement wayOrRelation, String tagName) {
+        if (wayOrRelation.hasTag(tagName)) {
+            return wayOrRelation.getTag(tagName);
         } else {
             return null;
         }
     }
 
+    public static Map<String, String> parseWayTags(ReaderWay ghReaderWay) {
+        Map<String, String> parsedWayTagValues = Maps.newHashMap();
+
+        // Parse street name, which is a concat of `name` and `ref` tags (if present)
+        parsedWayTagValues.put(OSM_NAME_TAG, getConcatNameFromOsmElement(ghReaderWay));
+
+        // Parse highway and direction tags, plus all tags needed for determining lane counts
+        for (String wayTag : ALL_TAGS_TO_PARSE) {
+            parsedWayTagValues.put(wayTag, getTagValueFromOsmElement(ghReaderWay, wayTag));
+        }
+
+        // Remove any tags that weren't present for this Way (ie the value was parsed as null)
+        parsedWayTagValues.values().removeIf(Objects::isNull);
+        return parsedWayTagValues;
+    }
+
     // if only `name` or only `ref` tag exist, return that. if both exist, return "<ref>, <name>". else, return null
     public static String getConcatNameFromOsmElement(ReaderElement wayOrRelation) {
-        String name = null;
-        if (wayOrRelation.hasTag("name")) {
-            name = wayOrRelation.getTag("name");
-        }
+        String name = getTagValueFromOsmElement(wayOrRelation, "name");
         if (wayOrRelation.hasTag("ref")) {
             name = name == null ? wayOrRelation.getTag("ref") : wayOrRelation.getTag("ref") + ", " + name;
         }
         return name;
     }
+
+    public static void updateOsmIdToWayTags(Map<Long, Map<String, String>> osmIdToWayTags, Long osmId, Map<String, String> newTagValues) {
+        Map<String, String> currentWayTags = new HashMap<>(osmIdToWayTags.getOrDefault(osmId, Maps.newHashMap()));
+        for (String tag : newTagValues.keySet()) {
+            if (!currentWayTags.containsKey(tag)) {
```

Comment:
should this check to ever fail? might be better to fail fast or at least log a warning if we should never get duplicate or conflicting values for the same way. looks like we only grab the name from the relation if no name was parsed from the way, so i wouldn't expect this check to be relevant there

Reply:
Good question, I think this is a bit of a holdover from me flip-flopping on how I wanted the relation tag parsing to work wrt overwriting data we've already parsed

I moved to fast-failing now in the case where the tag already has been parsed for the way and stored in `osmIdToWayTags`, and will make sure that when working on the follow-up PR that parses new data from the relation, the check we do [here](https://github.com/replicahq/graphhopper/pull/161/files#diff-221bd148393011a56e5bef015d107b765efac49201a7cb43ac9819f3702ce0fcR149) to ensure we don't try to parse data we've already stored stays in place (or equivalent logic)